bayes_opt/domain_reduction.py

```diff
@@ -96,46 +99,101 @@ def _update(self, target_space: TargetSpace) -> None:
             np.abs(self.current_d) * (self.gamma - self.eta)
 
         self.r = self.contraction_rate * self.r
+        
+    def _windowed_max(self, target_space: TargetSpace) -> np.array:
+        """Returns the parameters that produce the greatest target value within its bounds.
+        """
+        # extract the components we need from the target space
+        params = np.copy(target_space.params.T)
+        target = np.copy(target_space.target)
+        bounds = np.copy(target_space.bounds)
+        
+        # create a mask by checking each params against its bounds
+        mask = np.zeros_like(target)
+        for n, row in enumerate(params):
+            lower_bound = bounds[n, 0]
+            upper_bound = bounds[n, 1]
+            mask += (row <= lower_bound).astype(int)
+            mask += (row >= upper_bound).astype(int)
+        mask = mask<1
+
+        # apply the mask
+        params = params.T[mask]
+        targets = target[mask]
+
+        best_params = params[np.argmax(targets)]
+
+        return best_params
 
     def _trim(self, new_bounds: np.array, global_bounds: np.array) -> np.array:
-        """ Tests new_bounds to enforce global_bounds and minimum_window."""
-        # sort the order of the bounds for each parameter
+        """
+        Adjust the new_bounds and verify that they adhere to global_bounds and minimum_window.
+
+        Parameters:
+        -----------
+        new_bounds : np.array
+            The proposed new_bounds that (may) need adjustment.
+
+        global_bounds : np.array
+            The maximum allowable bounds for each parameter.
+
+        Returns:
+        --------
+        new_bounds : np.array
+            The adjusted bounds after enforcing constraints.
+        """
+
+        #sort bounds
         new_bounds = np.sort(new_bounds)
 
-        # check if the new bounds exceed the global bounds
+        # Validate each parameter's bounds against the global_bounds
         for i, pbounds in enumerate(new_bounds):
-            # check if lower bound exceeds range 
+            # If the lower bound of the parameter is outside the global bounds, reset the lower bound
             if (pbounds[0] < global_bounds[i, 0] or pbounds[0] > global_bounds[i, 1]):
                 pbounds[0] = global_bounds[i, 0]
-                warn("""Domain reduction warning:
-                    A parameter's lower bound has exceeded its global boundaries.
-                    The offensive boundary has been reset, but the optimizer may not converge.""")
-                
-            # check if upper bound exceeds range
+                warn("""Domain Reduction Warning:
+                    A parameter's lower bound has exceeded its global limit.
+                    The offensive boundary has been reset, but be cautious of optimizer convergence.""")
```

Comment:
`optimizer` generally refers to the bayesian optimizer, which should still converge to something, just slower compared to when the SDRT is used. This message might be confusing for users who could think that the entire optimization procedure has broken down. You could write, e.g.:

```suggestion
                    The offensive boundary has been reset, but the optimizer may proceed without domain reduction.""")
```
but not super attached to it and there's probably a better way to phrase that.

Reply:
I rewrote when the warnings were triggered to be more specific and reduce excessive warnings. Please see updated version.